drivers/ethernet/phy/phy_microchip_ksz8081.c

```diff
@@ -0,0 +1,460 @@
+	if (ret) {
+		return ret;
+	}
+
+	if (config->phy_iface == KSZ8081_RMII) {
+		ctrl2 |= PHY_MC_KSZ8081_CTRL2_REF_CLK_SEL;
+	} else {
+		ctrl2 &= ~PHY_MC_KSZ8081_CTRL2_REF_CLK_SEL;
+	}
+
+	ret = phy_mc_ksz8081_write(dev, PHY_MC_KSZ8081_CTRL2_REG, (uint32_t)ctrl2);
+	if (ret) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int phy_mc_ksz8081_cfg_link(const struct device *dev,
+					enum phy_link_speed speeds)
+{
+	const struct mc_ksz8081_config *config = dev->config;
+	struct mc_ksz8081_data *data = dev->data;
+	int ret;
+	uint32_t anar;
+
+	/* Lock mutex */
+	ret = k_mutex_lock(&data->mutex, K_FOREVER);
+	if (ret) {
+		LOG_ERR("PHY mutex lock error");
+		return ret;
+	}
+
+	/* We are going to reconfigure the phy, don't need to monitor until done */
+	k_work_cancel_delayable(&data->phy_monitor_work);
+
+	/* DT configurations */
+	ret = phy_mc_ksz8081_static_cfg(dev);
+	if (ret) {
+		return ret;
+	}
+
+	/* Read ANAR register to write back */
+	ret = phy_mc_ksz8081_read(dev, MII_ANAR, &anar);
+	if (ret) {
+		LOG_ERR("Error reading phy (%d) advertising register", config->addr);
+		return -EIO;
+	}
+
+	/* Setup advertising register */
+	if (speeds & LINK_FULL_100BASE_T) {
+		anar |= MII_ADVERTISE_100_FULL;
+	} else {
+		anar &= ~MII_ADVERTISE_100_FULL;
+	}
+	if (speeds & LINK_HALF_100BASE_T) {
+		anar |= MII_ADVERTISE_100_HALF;
+	} else {
+		anar &= ~MII_ADVERTISE_100_HALF;
+	}
+	if (speeds & LINK_FULL_10BASE_T) {
+		anar |= MII_ADVERTISE_10_FULL;
+	} else {
+		anar &= ~MII_ADVERTISE_10_FULL;
+	}
+	if (speeds & LINK_HALF_10BASE_T) {
+		anar |= MII_ADVERTISE_10_HALF;
+	} else {
+		anar &= ~MII_ADVERTISE_10_HALF;
+	}
+
+	/* Write capabilities to advertising register */
+	ret = phy_mc_ksz8081_write(dev, MII_ANAR, anar);
+	if (ret) {
+		LOG_ERR("Error writing phy (%d) advertising register", config->addr);
+		return ret;
+	}
+
+	/* Unlock mutex */
+	ret = k_mutex_unlock(&data->mutex);
+	if (ret) {
+		LOG_ERR("PHY mutex unlock error");
+		return ret;
+	}
+
+	/* (re)do autonegotiation */
+	ret = phy_mc_ksz8081_autonegotiate(dev);
+	if (ret) {
+		LOG_ERR("Error in autonegotiation");
+		return ret;
+	}
+
+	/* Get link status */
+	ret = phy_mc_ksz8081_get_link(dev, &data->state);
+	if (ret) {
+		return ret;
+	}
+
+	/* Start monitoring */
+	k_work_schedule(&data->phy_monitor_work,
+				K_MSEC(CONFIG_PHY_MONITOR_PERIOD));
+
+	/* Log the results of the configuration */
+	LOG_INF("PHY %d is %s", config->addr, data->state.is_up ? "up" : "down");
+	LOG_INF("PHY (%d) Link speed %s Mb, %s duplex\n", config->addr,
+		(PHY_LINK_IS_SPEED_100M(data->state.speed) ? "100" : "10"),
+		PHY_LINK_IS_FULL_DUPLEX(data->state.speed) ? "full" : "half");
+
+	return 0;
+}
+
+static int phy_mc_ksz8081_link_cb_set(const struct device *dev,
+					phy_callback_t cb, void *user_data)
+{
+	struct mc_ksz8081_data *data = dev->data;
+
+	data->cb = cb;
+	data->cb_data = user_data;
+
+	phy_mc_ksz8081_get_link(dev, &data->state);
+
+	data->cb(dev, &data->state, data->cb_data);
```

Comment:
A bus fault occurs when `data->cb` is called here since `nxp_enet_phy_cb` calls `net_eth_carrier_on `which relies on `data->iface`. However iface seems to be not initialized yet. Default CONFIG_ETH_INIT_PRIORITY=80 and default CONFIG_NET_INIT_PRIO=90. Shouldn't we check if `data->iface` is initialized before calling `data->cb`?

Reply:
thanks, fixed